File: application/backend-credit-card-enrollment/backend-c#/CreditCard/Enrollment/Projection/EnrollmentList/ProductNameRepository.cs
using CreditCardEnrollment.Common.Projection;
using MongoDB.Driver;

namespace CreditCardEnrollment.CreditCard.Enrollment.Projection.EnrollmentList;

public class ProductNameRepository {
    private readonly MongoTransactionalProjectionOperator _mongoOperator;

    public ProductNameRepository(MongoTransactionalProjectionOperator mongoOperator) {
        _mongoOperator = mongoOperator;
    }

    public void Save(ProductName productName) {
        Collection().ReplaceOne(
            p => p.Id == productName.Id,
            productName,
            new ReplaceOptions { IsUpsert = true }
        );
    }

    public ProductName? FindOneById(string id) {
        return Collection().Find(p => p.Id == id).FirstOrDefault();
    }

    private IMongoCollection<ProductName> Collection() {
        return _mongoOperator.Operate().GetCollection<ProductName>("CreditCard_Enrollment_ProductName");
    }
}

File: application/backend-credit-card-enrollment/backend-typescript/src/creditCard/enrollment/projection/enrollmentList/ProductNameRepository.ts
import {ProductName} from "./ProductName";
import {MongoTransactionalProjectionOperator} from "../../../../common/projection/MongoTransactionalProjectionOperator";

export class ProductNameRepository {
    private readonly collectionName = 'CreditCard_Enrollment_ProductName';

    constructor(private readonly mongoOperator: MongoTransactionalProjectionOperator) {}

    async save(productName: ProductName): Promise<void> {
        await this.mongoOperator.replaceOne(
            this.collectionName,
            { _id: productName._id },
            productName,
            { upsert: true }
        );
    }

    async findOneById(_id: string): Promise<ProductName | null> {
        const results = await this.mongoOperator.find<ProductName>(this.collectionName, { _id });
        return results.length > 0 ? results[0] : null;
    }
}